data/tutorials/lg_08_error_handling.md

```diff
@@ -125,67 +191,549 @@ let () =
+  second parameter).
+* Lines 2 and 3: same as in the original
+* Line 4: `try` or `match` is removed
+* Line 5: `String.sub` is applied, if the previous step didn't fail, otherwise
+  the error is forwarded
+* Line 6: if nothing was found earlier, and if isn't empty, `fqdn` is returned
+  as a fallback
+
+When used to handle errors with catch statements, it requires some time to get
+used the latter style. The key idea is avoiding or deferring from directly
+looking into option values. Instead, pass them along using _ad-hoc_ pipes (such
+as `map` and `bind`). Erik Meijer calls that following the happy path. Visually,
+it also resembles the “early return“ pattern often found in C.
+
+One of the limitations of the option type is it doesn't record the reason which
+prevented having a return value. `None` is silent, it doesn't say anything about
+what went wrong. For this reason, functions returning option values should
+document the circumstances under which it may return `None`. Such a
+documentation is likely to ressemble to the one required for exceptions using
+`@raise`. The `Result` type is intended to fill this gap: manage error in data,
+like option values, but also provide information on errors, like exceptions. It
+is the topic of the next section.
+
+## `Result` Type
+
+The `Result` module of the standard library defines the following type:
 
 ```ocaml
 type ('a, 'b) result =
   | Ok of 'a
   | Error of 'b
 ```
 
-A value `Ok x` means that the computation succeeded with `x`, and
-a value `Error e` means that it failed.
-Pattern matching can be used to deal with both cases, as with any
-other sum type. The advantage here is that a function `a -> b` that
-fails can be modified so its type is `a -> (b, error) result`,
-which makes the failure explicit.
-The error case `e` in `Error e` can be of any type
-(the `'b` type variable), but a few possible choices
-are:
+A value `Ok x` means that the computation succeeded and produced `x`, and a
+value `Error e` means that it failed and `e` represents whatever error
+information has been collected in the process. Pattern matching can be used to
+deal with both cases, as with any other sum type. However using `map` and `bind`
+can be more convenient, maybe even more as it was with `Option`.
+
+Before taking a look at `Result.map`, let's think about `List.map` and
+`Option.map` under a changed perspective. Both functions behave as identity when
+applied to `[]` or `None`, respectively. That's the only possibility since those
+parameters don't carry any data. Which isn't the case in `Result` with its
+`Error` constructor. Nethertheless, `Result.map` is implemented likewise, on
+`Error`, it also behaves like identity.
+```ocaml
+let map f = function
+| Ok x -> Ok (f x)
+| Error e -> Error e
+```
+
+The `Result` module has two map functions: the one we've just seen and another
+one, with the same logic, applied to `Error`:
+```ocaml
+let map_error f = function
+| Ok x -> Ok x
+| Error e -> f e
```

Comment:
If I add the `;;` at the end, I get: 
`val map_error : ('a -> ('b, 'c) result) -> ('b, 'a) result -> ('b, 'c) result =
  <fun>` 
Should this be included?